### PR TITLE
Applied

### LINKED ISSUE #11: Revamp Mul, Ldiv, and Inv as Applied
A more general approach is to support general application of functions. The current lazy operations could be type aliases.

### PR DESCRIPTION
**Pull request by dlfivefifty**
This adds `Applied`, which will fix #11 and #14.

### CODE REVIEW

**Comment by dlfivefifty** [comment]
@tkf Could you have a look and say if you have any suggestions for the changes made to `Add`. I've made it an `Applied` which makes it consistent with `Mul`.

File: test/multests.jl [hunk 1]
```diff
@@ -677,8 +675,8 @@ end
 
 @testset "Add" begin
     @testset "gemv Float64" begin
-        for A in (Add(randn(5,5), randn(5,5)),
-                  Add(randn(5,5), view(randn(9, 5), 1:2:9, :))),
+        for A in (AddArray(randn(5,5), randn(5,5)),
+                  AddArray(randn(5,5), view(randn(9, 5), 1:2:9, :))),
```
[the remark this answers is not in the document]
Comment (dlfivefifty) [reply test/multests.jl]: Yes good point. This needs a redesign of the code, though it's touching against a broader point: given a general `applied` "tree" of operations, how's best to simplify it? For example, in this case what we actually want is
```julia
applied(*, applied(+, A, B,C) , D)
``` 
to be "simplified" to 
```julia
applied(+, applied(*, A, D), applied(*, B, D), applied(*,C,D))
```
Another such case is choosing multiplication orders to minimize cost.

[the remark this answers is not in the document]
Comment (tkf) [reply test/multests.jl]: I'd suggest to store the tree of operations as-is and do the simplification/transformation inside `materialize`.  We can then define "optimization passes" to look at the tree and turn it into more efficient one (maybe in the `instantiate` phase?).  If there is no transformation at "tree build time" then it's straight-forward to implement your own case-specific `materialize` function for `ApplyArray` or `Applied`.

Or, maybe you are wondering how `ApplyArray` and `Applied` interact?  I think it's better to always use `ApplyArray` in the outer most tree.  That is to say,

```julia
ApplyArray(applied(*, ApplyArray(applied(+, A, B, C)) , D))
```

is "canonicalized" to

```julia
ApplyArray(applied(*, applied(+, A, B, C) , D))
```

at the tree build time.  And, repeating my point in the first paragraph, it then is turned into

```julia
applied(+, applied(*, A, D), applied(*, B, D), applied(*, C, D))
```

inside `materialize` later.

[the remark this answers is not in the document]
Comment (dlfivefifty) [reply test/multests.jl]: At this point in time I think we should only support 1 version, which should be `applied`. 

Though your fast implementation of `(A+B+C)*D` is arguably premature to include, without having a well-defined "optimization" infrastructure. Is this feature crucial to you, or could that part be moved to your own code for the time being?

[the remark this answers is not in the document]
Comment (dlfivefifty) [reply test/multests.jl]: PS Arguably the lowering of `c .= Mul(A,b)` to `materialize!(MulAdd(1.0, A, b, 0.0, c))` also fits the pattern of a "simplifying" operation. Probably `MulAdd` should be type-aliased to `applied(+, applied(*, α, A, b), applied(*, β, c))`

**Comment by zsoerenm** [comment]
Great work!
Are there plans to optimize `Mul(A, A')` to `BLAS.syrk('U', 'N', 1, A)`?

I often have weighting matrix `W` too, which is either diagonal or symmetric. Unfortunately there is BLAS equivalent for `Mul(A, W, A')`. But could I force the output to be symmetric something like `B = Symmetric(Matrix{Float64}(undef, 10, 10)); B .= Mul(A, W, A')`?

**Comment by dlfivefifty** [comment]
Yes, supporting custom matrix types is a big motivation (especially when they live on the GPU or parallel memory). The syntax you asked for should eventually work, though also possible is the shorthand ` SymmetricMatrix(Mul(A,W,A’)) `.

An example of this already in action is converting ` MulArray ` of banded matrices to a ` BandedMatrix `.

**Pull request merged**